error.go

```diff
@@ -23,26 +23,47 @@ import (
 	"strings"
 
 	"google.golang.org/protobuf/proto"
-	"google.golang.org/protobuf/reflect/protoreflect"
 	"google.golang.org/protobuf/types/known/anypb"
 )
 
-const commonErrorsURL = "https://connect.build/docs/go/common-errors"
+const (
+	commonErrorsURL          = "https://connect.build/docs/go/common-errors"
+	defaultAnyResolverPrefix = "type.googleapis.com/"
+)
 
 // An ErrorDetail is a self-describing Protobuf message attached to an *Error.
 // Error details are sent over the network to clients, which can then work with
-// strongly-typed data rather than trying to parse a complex error message.
+// strongly-typed data rather than trying to parse a complex error message. For
+// example, you might use details to send a localized error message or retry
+// parameters to the client.
 //
-// The ErrorDetail interface is implemented by Protobuf's Any type, provided in
-// Go by the google.golang.org/protobuf/types/known/anypb package. The
-// google.golang.org/genproto/googleapis/rpc/errdetails package contains a
-// variety of Protobuf messages commonly wrapped in anypb.Any and used as error
-// details.
-type ErrorDetail interface {
-	proto.Message
+// The google.golang.org/genproto/googleapis/rpc/errdetails package contains a
+// variety of Protobuf messages commonly used as error details.
+type ErrorDetail struct {
+	pb       *anypb.Any
+	wireJSON string // preserve human-readable JSON
+}
+
+// NewErrorDetail constructs a new error detail.
+func NewErrorDetail(msg proto.Message) (*ErrorDetail, error) {
+	pb, err := anypb.New(msg)
+	if err != nil {
+		return nil, err
+	}
+	return &ErrorDetail{pb: pb}, nil
+}
+
+// Type is the fully-qualified name of the Detail's Protobuf message (for
+// example, acme.foo.v1.FooDetail).
+func (d *ErrorDetail) Type() string {
+	return strings.TrimPrefix(d.pb.TypeUrl, defaultAnyResolverPrefix)
```

Comment:
Can you add a comment as to why this is trimmed? I can't exactly remember and might be helpful for people in the future

Reply:
Sure, can do.